src/navigation/navigation-events.js

```diff
@@ -125,29 +125,39 @@ if (isInBrowser) {
     return originalRemoveEventListener.apply(this, arguments);
   };
 
-  window.history.pushState = patchedUpdateState(window.history.pushState);
-  window.history.replaceState = patchedUpdateState(window.history.replaceState);
-
-  function patchedUpdateState(updateState) {
+  window.history.pushState = patchedUpdateState(
+    window.history.pushState,
+    "pushState"
+  );
+  window.history.replaceState = patchedUpdateState(
+    window.history.replaceState,
+    "replaceState"
+  );
+
+  function patchedUpdateState(updateState, methodName) {
     return function () {
       const urlBefore = window.location.href;
       const result = updateState.apply(this, arguments);
       const urlAfter = window.location.href;
 
       if (!urlRerouteOnly || urlBefore !== urlAfter) {
-        urlReroute(createPopStateEvent(window.history.state));
+        urlReroute(createPopStateEvent(window.history.state, methodName));
       }
 
       return result;
     };
   }
 
-  function createPopStateEvent(state) {
+  function createPopStateEvent(state, originalMethodName) {
     // https://github.com/single-spa/single-spa/issues/224 and https://github.com/single-spa/single-spa-angular/issues/49
     // We need a popstate event even though the browser doesn't do one by default when you call replaceState, so that
-    // all the applications can reroute.
+    // all the applications can reroute. We explicitly identify this extraneous event by setting singleSpa=true and
+    // originalMethodName=<pushState|replaceState> on the event instance.
     try {
-      return new PopStateEvent("popstate", { state });
+      const evt = new PopStateEvent("popstate", { state });
+      evt.singleSpa = true;
+      evt.originalMethodName = originalMethodName;
```

Comment:
Hmm naming this property is pretty tricky. I'd like something that clearly indicates this is not part of the browser spec and is specific to single-spa. Even if it makes the name a bit cumbersome. Maybe `singleSpaOriginalMethodName`, `singleSpaCause`, `singleSpaTrigger`, `singleSpaStateChange`???

Reply:
I'm onboard with that. `singleSpaTrigger` sounds best to me.

WDYT about a helpful message in `evt.singleSpa` property? Is it inappropriate to use `formatErrorMessage` here to help point others to a help article?

`evt.singleSpa = 'Extraneous popstate event dispatched by SingleSpa, triggered by push/replace state.';`

OR

`evt.singleSpa = formatErrorMessage(123, 'Synthetic popstate event.');`

Reply:
I think I'd prefer just to keep  `evt.singleSpa` a boolean, and add documentation about this behavior to our API docs for those that search for it.